protocols/dcutr/src/handler/relayed.rs

```diff
@@ -286,38 +233,33 @@ impl ConnectionHandler for Handler {
 
     fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
         match event {
-            Command::Connect { obs_addrs } => {
+            Command::Connect => {
                 self.queued_events
                     .push_back(ConnectionHandlerEvent::OutboundSubstreamRequest {
                         protocol: SubstreamProtocol::new(
-                            protocol::outbound::Upgrade::new(obs_addrs),
+                            protocol::outbound::Upgrade::new(self.holepunch_candidates.clone()),
                             (),
                         ),
                     });
-            }
-            Command::AcceptInboundConnect {
-                inbound_connect,
-                obs_addrs,
-            } => {
-                if self
-                    .inbound_connect
-                    .replace(inbound_connect.accept(obs_addrs).boxed())
-                    .is_some()
-                {
-                    log::warn!(
-                        "New inbound connect stream while still upgrading previous one. \
-                         Replacing previous with new.",
-                    );
-                }
-            }
-            Command::UpgradeFinishedDontKeepAlive => {
-                self.keep_alive = KeepAlive::No;
+                self.attempts += 1;
             }
         }
     }
 
     fn connection_keep_alive(&self) -> KeepAlive {
-        self.keep_alive
+        if !self.queued_events.is_empty() {
+            return KeepAlive::Yes;
+        }
+
+        if self.inbound_connect.is_some() {
+            return KeepAlive::Yes;
+        }
+
+        if self.attempts < MAX_NUMBER_OF_UPGRADE_ATTEMPTS {
```

Comment:
Instead of referencing the constant here, I think it would be nicer if we passed the value in at the constructor.

That allows us to keep the constant an implementation detail of the behaviour. What do you think guys?

Reply:
Passing it in would make the constant mutable, right? In other words, there is no way to define a constant in a `struct` in Rust. I find a global `const` easier to reason about than a maybe-const inside a `struct`.

Reply:
It is mutable as a field then yes. At the moment though, the separation between behaviour and handler isn't very clean.

The handler trusts that the behaviour will operate on the same constant, it implicitly depends on it.

If we pass it in, the behaviour explicitly initializes the handler with this value, thus removing the dependency on the global const.

It makes the handler more cohesive which I'd argue is easier to understand. Mutating a variable that says "limit" would be a local bug that is pretty easy to find. Plus, it would be easy to write a unit test for this behaviour (not saying we need one).

Reply:
Agreed with the logic here, since it is not possible to declare a constant inside a Rust structure, the best we could do is to "simulate" the behavior of a constant. So it's better for me to declare a real global constant.